src/main/java/org/forgerock/openam/auth/nodes/SetProfilePropertyNode.java

```diff
@@ -90,41 +112,63 @@ public SetProfilePropertyNode(@Assisted Config config, CoreWrapper coreWrapper)
     }
 
     @Override
-    public Action process(TreeContext context) {
+    public Action process(TreeContext context) throws NodeProcessException {
         String username = context.sharedState.get(USERNAME).asString();
         String realm = context.sharedState.get(REALM).asString();
         AMIdentity userIdentity = coreWrapper.getIdentity(username, realm);
 
-        Map<String, Set<String>> attributes = new HashMap<>();
-        for (Map.Entry<String, String> entry : config.properties().entrySet()) {
-            String key = entry.getKey();
-            String propertyValue = entry.getValue();
-
-            String result = null;
-            if (propertyValue.startsWith("\"")) {
-                result = propertyValue.substring(1, propertyValue.length() - 1);
-            } else if (context.sharedState.isDefined(propertyValue)) {
-                result = context.sharedState.get(propertyValue).asString();
-            }
-
-            if (StringUtils.isNotEmpty(result)) {
-                attributes.put(key, singleton(result));
-            }
-        }
-
-        config.transientProperties().forEach((key, value) -> {
-            if (context.transientState.isDefined(value)) {
-                attributes.put(key, singleton(context.transientState.get(value).asString()));
-            }
-        });
+        Map<String, Set<String>> attributes = createAttributes(context.transientState, userIdentity,
+                                                               config.transientProperties(),
+                                                               createAttributes(context.sharedState, userIdentity,
+                                                                                config.properties(), new HashMap<>()));
 
         try {
             userIdentity.setAttributes(attributes);
             userIdentity.store();
         } catch (IdRepoException | SSOException ex) {
-            LOGGER.error("Unable to update user {} in realm {} with attributes {}", username, realm, attributes, ex);
+            LOGGER.error("Unable to update user " + username + " in realm " + realm + " with attributes " + attributes,
+                         ex);
         }
 
         return goToNext().build();
     }
+
+    private Map<String, Set<String>> createAttributes(JsonValue state, AMIdentity userIdentity,
```

Comment:
May I suggest to change this to a non-chained version? The attributes parameter is used as a return value essentially, and it isn't really checked if it already has a mapping. I think it would result in an easier to read code.